### DISCUSSION TITLE
Android Launch Profile + `environmentVariables` (launchSettings.json)

### DISCUSSION THREAD

**Discussion by Mike-E-angelo** [unoplatform/uno#19133] [Q&A] [answered]
I have been using launchSettings.json to assign `environmentVariables` (most notably `DOTNET_ENVIRONMENT`) when launching different platforms.  When launching the Android emulator this appears to be ignored. 🤔

I've searched around and found [this StackOverflow question](https://stackoverflow.com/questions/78880912/is-it-possible-to-set-the-framework-in-launchsettings-json) as the only somewhat-relevant resource of interest around this topic.

Is there guidance on how to launch the Android emulator so that it picks up values and configurations from a launchSettings.json profile?  Thank you for any assistance you can provide.

**Comment by MartinZikmund** [comment]
Are you able to do this with a plain .NET Android/.NET MAUI app? If not, I think it is better suited as a question on the Microsoft .NET Android repository https://github.com/dotnet/android

**Reply by Mike-E-angelo** [reply]
Hi @MartinZikmund thank you for your reply. 🙏  I have no experience with Android/Maui and only Uno at the moment. :)  Are you saying that you do not use `environmentVariables` from `launchSetttings.json` in your projects?  If you have a suggestion on how to assign this variable outside of using a compiler directive I would be interested in any suggestions.  Thank you again for your assistance.

**Reply by MartinZikmund** [reply]
It would probably be enough to just try creating a blank MAUI and see if the environment variable is applied or not - my point was that as long setting the variable in `launchSettings.json` does not have an effect for plain .NET Android app, it explains why it does not have an effect in case of Uno Platform either.

As an alternative you could add a custom a compiler constant that you would set and based on it do some things conditionally (see https://stackoverflow.com/questions/5049323/msbuild-with-solution-files-with-defineconstants for a simple example)

**Reply by Mike-E-angelo** [reply]
Ah that's the thing @MartinZikmund, there isn't a profile for the Android emulator in `launchSettings.json`.  The profiles found there are everything but Android as far as I can tell.  In my case I tried to add it manually but did not have any luck with that so that ended up with searching + asking here.

Thank you for the suggestion for compiler constants but as mentioned earlier I am really trying to stay away from compiler directives as I consider them an anti-pattern that negatively impacts code quality.  And yes, I realize this is a minority opinion. 😇

As a final add here, I tried the following in the csproj msbuild file but it did not do anything:
```
<PropertyGroup Condition="'$(Configuration)'=='Debug' and '$(TargetFramework)'=='net9.0-android'">
    <EnvironmentName>Development</EnvironmentName>
</PropertyGroup>
```

Having something like this would be valuable.  I will continue to search and ask on Android-related forums and if I encounter a good answer I will be sure to update here. 👍🙏

**Reply by jeromelaban** [reply]
Environment variables can be specified in the `environment.conf` file, not in the csproj. See this file in [uno app](https://github.com/unoplatform/uno.templates/blob/993a91104c3620545fc16addd84db2a9dc4cbec4/src/Uno.Templates/content/unoapp/MyExtensionsApp.1/Platforms/Android/environment.conf) templates.

**Reply by Mike-E-angelo** [reply]
Ah I do see that, @jeromelaban thank you for pointing that out and I do see this in my solution as it was generated from that template. 👍

However, if I add `DOTNET_ENVIRONMENT=Development` to this, that will also assign this value when the application is deployed to production/testing/etc and therefore with an incorrect value, correct? 🤔

**Reply by jeromelaban** [reply, marked as answer]
> However, if I add DOTNET_ENVIRONMENT=Development to this, that will also assign this value when the application is deployed to production/testing/etc and therefore with an incorrect value, correct? 🤔

That is correct. You can have two different files, one per environment, if you add multiple files as `AndroidEnvironment` they will be concatenated to the other ones. `environment.conf` is [automatically added by Uno when present](https://github.com/unoplatform/uno/blob/8d836a8309139a87431e4d4eb9d5879f2fc325ea/src/Uno.Sdk/targets/Uno.SingleProject.Android.targets#L26) but you can add you own in your csproj.

**Reply by Mike-E-angelo** [reply]
Woohoo, awesome!  Thank you for your guidance and patience as I stand on my feet here @jeromelaban, @MartinZikmund, and the team.  It is appreciated. 🙏

For thoroughness, here is the MsBuild element that I used:

```
<ItemGroup Condition="'$(TargetFramework)'=='net9.0-android'">
	<AndroidEnvironment Include="Platforms/Android/environment.$(Configuration).conf"
	                    Condition="Exists('Platforms/Android/environment.$(Configuration).conf')"
	                    Link="$([MSBuild]::MakeRelative ('$(MSBuildProjectDirectory)', '%(Identity)'))"
	                    IsDefaultItem="true" />
</ItemGroup>
```

I then added `Platforms/Android/environment.debug.conf` in my Uno application project with the following contents:

```
# See this for more details: http://developer.xamarin.com/guides/android/advanced_topics/garbage_collection/
MONO_GC_PARAMS=bridge-implementation=tarjan,nursery-size=32m,soft-heap-limit=256m
DOTNET_ENVIRONMENT=Development
```
